Webapp/sources/services/utils.py

```diff
@@ -1,5 +1,46 @@
+import contextlib
+import os
+import re
+from typing import Tuple
+
+import toml
+
+
 def camelCase_to_snake_case(string: str) -> str:
     """
     Converts a string from camelCase which is often used in JavaScript to the preferred snake_case.
     """
     return "".join(["_" + c.lower() if c.isupper() else c for c in string]).lstrip("_")
+
+
+def get_app_version() -> Tuple[str, str]:
+    app_directory = os.path.dirname(os.path.abspath(__file__))
+
+    # Traverse up directories to reach the project root
+    project_root = os.path.abspath(os.path.join(app_directory, "..", "..", "..", ".."))
+
+    # Construct paths relative to the project root
+    pyproject_path = os.path.join(project_root, "pyproject.toml")
+    hash_file_path = os.path.join(project_root, "hash.txt")
```

Comment:
I've just had a thought. These project variables could instead be defined in your `config.py` and then imported into your various apps